test/system/src/test/java/io/pravega/test/system/AbstractFailoverTests.java

```diff
@@ -97,22 +94,90 @@
         final CompletableFuture<Void> newWritersComplete = new CompletableFuture<>();
         final CompletableFuture<Void> readersComplete = new CompletableFuture<>();
         final List<CompletableFuture<Void>> txnStatusFutureList = synchronizedList(new ArrayList<>());
-        final AtomicBoolean txnWrite = new AtomicBoolean(false);
+
+        final ConcurrentSet<UUID> committingTxn = new ConcurrentSet<>();
+        final ConcurrentSet<UUID> abortedTxn = new ConcurrentSet<>();
+        final AtomicLong eventData = new AtomicLong();
+        final boolean txnWrite;
+
+        private final ConcurrentHashMap<Long, Integer> eventMap = new ConcurrentHashMap<>();
+
+        TestState(boolean txnWrite) {
+            this.txnWrite = txnWrite;
+        }
+
+        void eventWritten(Long event) {
+            eventMap.putIfAbsent(event, 0);
+        }
+
+        void eventRead(Long event) {
+            eventMap.compute(event, (x, y) -> {
+                if (y == null) {
+                    return 1;
+                } else {
+                    return y + 1;
+                }
+            });
+        }
+
+        Long getNext() {
+            return eventData.getAndIncrement();
+        }
+
+        int getEventWrittenCount() {
+            return eventMap.size();
+        }
+
+        int getEventReadCount() {
+            return eventMap.values().stream().reduce(0, (x, y) -> x + y);
+        }
+
+        void printAnomalies() {
+            List<Long> notRead = eventMap.entrySet().stream().filter(x -> x.getValue() == 0)
+                    .map(Map.Entry::getKey).collect(Collectors.toList());
+            if (notRead.size() > 0) {
+                log.error("anomalies, unread events => {}", notRead);
+            }
+
+            Map<Long, Integer> duplicates = eventMap.entrySet().stream().filter(x -> x.getValue() > 1)
+                    .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+            if (duplicates.size() > 0) {
+                log.error("anomalies, duplicate events with count => {}", duplicates);
```

Comment:
Also, can you please capitalise the first letter of the log statement, here and everywhere